### ISSUE TITLE
Issue with [prepend] and [append] attributes doesn't work.

### ISSUE THREAD

**Issue by Shin-Ogata** [i18next/jquery-i18next#20]
Thank you for your great job.

I found problem for the [prepend] and [append] attributes doesn't work in version 1.0.1.
My HTML is following and this works fine in version 0.2.0.

```
<input id="some-checkbox" type="checkbox" name="some-checkbox-group" value="hoge" checked="checked">
<label for="some-checkbox" data-i18n="[prepend]foo.bar">
  <span class="cool-effect"></span>
  <span class="cool-indicator"></span>
</label>
```

Because in localze process, jquery causes exception by "**#[prepend]foo.bar**" selector syntax error.
(invalid trimmedKey)
https://github.com/i18next/jquery-i18next/blob/master/src/index.js#L43

I know that this change depends on #19  proposal.
But I think, the proposal is so far in the different function.
Recently, HTML tag is used to make the design rich, too.
Therefore, should the library also provides to not changing HTML structure like before?

**Comment by jamuhl** [comment]
not sure why that should throw an exception? `ele.find('#' + trimmedKey).length`should be an empty array or not? did not test that to be honest...personally i don't use jquery for years now.

but agree...PR would be very welcome...

**Comment by Shin-Ogata** [comment]
The error is caused by **jquery.find()** taking invalid selector **#[prepend]foo.bar**.
**#[prepend]** is invalid ID SELECTOR format.
jquery.find() does not finish and throws an error.

Call stack is below.(I use jquery 2.2.4)
- call stack ↓

https://github.com/i18next/jquery-i18next/blob/master/src/index.js#L43
https://github.com/jquery/jquery/blob/2.2.4/dist/jquery.js#L2792
https://github.com/jquery/jquery/blob/2.2.4/dist/jquery.js#L2683
https://github.com/jquery/jquery/blob/2.2.4/dist/jquery.js#L838
https://github.com/jquery/jquery/blob/2.2.4/dist/jquery.js#L2063
https://github.com/jquery/jquery/blob/2.2.4/dist/jquery.js#L2125
https://github.com/jquery/jquery/blob/2.2.4/dist/jquery.js#L1468

↑ error thrown

I think the best solution is a rollback of change in #19.
Because #19 issue is handling the word order and I think it should use **data-i18n-options**.
https://github.com/i18next/jquery-i18next#using-options-in-translation-function

If you'd like also to support #19 condition, to prepare new keywords and do as follows separately would be good.
ex) prepend-with-dom, append-with-dom

usage:

```
<a id="btn1" href="#" data-i18n="[prepend-with-dom]key.for.title" data-i18n-target="#selector">insert before me, please!</a>
$("#btn1").localize();
```

source:

```
:
      if (attr === 'html') {
:
      } else if (attr === 'prepend') {
        ele.prepend(i18next.t(key, extendDefault(opts, ele.html())));
      } else if (attr === 'append') {
        ele.append(i18next.t(key, extendDefault(opts, ele.html())));
      } else if (attr === 'prepend-with-dom') {
        var selector = ele.data(options.targetAttr); // data-i18n-target
        if (selector && ele.find(selector).length > 0) {
          ele.find(selector).html(i18next.t(key, extendDefault(opts, ele.html())));
        } else if (selector && selector[0] === '#') {
          ele.prepend('<span id=' + selector.replace('#', '') + '>' + i18next.t(key, extendDefault(opts, ele.html())) + '</span>');
        } else if (selector && selector[0] === '.') {
          ele.prepend('<span class=' + selector.replace('.', '') + '>' + i18next.t(key, extendDefault(opts, ele.html())) + '</span>');
        }
      } else if (attr === 'append-with-dom') {
        var selector = ele.data(options.targetAttr); // data-i18n-target
        if (selector && ele.find(selector).length > 0) {
          ele.find(selector).html(i18next.t(key, extendDefault(opts, ele.html())));
        } else if (selector && selector[0] === '#') {
          ele.append('<span id=' + selector.replace('#', '') + '>' + i18next.t(key, extendDefault(opts, ele.html())) + '</span>');
        } else if (selector && selector[0] === '.') {
          ele.append('<span class=' + selector.replace('.', '') + '>' + i18next.t(key, extendDefault(opts, ele.html())) + '</span>');
        }
      } else if (attr.indexOf('data-') === 0) {
:
```

It seem to be able to write a better code, 
and it is difficult for me to consider good key words.

Best regards,

**Comment by jamuhl** [comment]
#19 addresses language changes which do not trigger a reload - the html already got something prepended as we do not have a clue that we already run localize in one language and prepended something the second time we prepend we prepend infront of the first translation instead of replacing it.

was after en: `english value are prepended in front of this`
actual result after change lng to de: `german value english value are prepended in front of this`
expected result after change lng to de: `german value are prepended in front of this` 

so result should be generating a valid id i think @bhaskerchari what you think?

**Comment by bhaskerchari** [comment]
Hi @jamuhl ,

The fix for that issue was to use the below line of code:

Actually I found the issue the day after I posted the fix for the issue [(https://github.com/i18next/jquery-i18next/issues/19)] but was busy to post the fix.

Please update the lines of code in jquery-i18next.js with below 2 lines of code(line no's : 47 and 48):

// Removing any spaces and any invalid special characters in key for valid html5 id  
var trimmedKey = key.replace(/[^a-z0-9-_:.]|^[^a-z]+/gi, "");

Thanks,
Bhasker

**Comment by Shin-Ogata** [comment]
I think #19 fix should be returned to the beginning because the feature is no longer support range of the library.

You're going to make the **VALID ID** and to change the html structure.
I think that's a mistake.

For example, following case generates invalid html. (**duplicate id** generated.)

```
<div data-i18n="[prepend]information"> is wealth.</div>
<div data-i18n="[prepend]information"> is a weapon.</div>
```

↓

```
<div data-i18n="[prepend]information"><span id="#information">GE - Information</span> is wealth.</div>
<div data-i18n="[prepend]information"><span id="#information">GE - Information</span> a weapon.</div>
```

The job to manage the structure of the HTML should not be this library's one but the client's one.
In the case of #19, you should use the following HTML instead of "[prepend]" function.

```
<div><span data-i18n="information"></span> is wealth.</div>
```

So, the BEST solution is a rollback of change in #19.

Regards,

**Comment by jamuhl** [comment]
i agree with @Shin-Ogata will rollback the changes as the points seem valid and a sample is provided to solve the demanded problem.

**Comment by bhaskerchari** [comment]
Hello @Shin-Ogata ,

I see your point but you are not getting the sense of [https://github.com/i18next/jquery-i18next/issues/19](url) and the rollback of the change doesn't help it.

Let me explain this to you:

`<div data-i18n="[prepend]information"> is wealth.</div>`

If I switch the language multiple times it keeps prepending the same text again and again when you switch the language multiple times and localize it.

For example if I switch the language using language switcher 4 times then the output would be something like below (considering 2 languages English and Russian) which is not the expected behaviour:

output:
`Information Информация Information Информация Information is wealth.`

The way  you suggested helps which I am aware of but this way I have to change my DOM structure manually wherever I have such usecase which is kind of redundant as [prepend] and [append] attributes are meant for that purpose and it doesn't work as expected with the rollback of changes.

`<div><span data-i18n="information"></span> is wealth.</div>`

For the problem which you talked about duplicate ids, even though it might generate duplicate id's it won't have any problem because lib tries to find the element with id on the respective node element so it will only find one correct matching element.

Thanks,
Bhasker

**Comment by bhaskerchari** [comment]
@jamuhl ,

If you agree with @Shin-Ogata then it doesn't make any sense in having the [prepend] and [append] attributes and are basically useless and should be removed from the implementation feature as part of the library because it doesn't work as intended as explained above.

Thanks,
Bhasker

**Comment by jamuhl** [comment]
hm...really depends. basically this was never meant to build spa's using jquery. what i thought of was a language change triggers a page reload -> which obviously completely avoid such situations where we run over already translated content again.

but honestly i'm unsure...both of you got a valid point. And as for over 5 years no one complained about it being of no sense i think rolling back might be the proper solution. More opinions on this would help but right now it's 1 vs. 1 - and time ticks as i need to fix the broken version asap.

**Comment by Shin-Ogata** [comment]
Hi @bhaskerchari,

I have been using the [prepend] and [append] functions follow situation from i18next ver1.x.
- if [prepend] is not used

```
<div data-i18n="foo.bar">
  <!-- expected localized.foo.bar text here -->
  <div class="special-design"></div>
</div>
```

↓

```
<div data-i18n="foo.bar">
  localized.foo.bar.text.here 
</div>
```

muuu, **<div class="special-design"></div>** is trimmed...... _|￣|○
- if [prepend] is used

```
<div data-i18n="[prepend]foo.bar">
  <!-- expected localized.foo.bar text here -->
  <div class="special-design"></div>
</div>
```

↓

```
<div data-i18n="foo.bar">
  localized.foo.bar.text
  <!-- expected localized.foo.bar text here -->
  <div class="special-design"></div>
</div>
```

everything fine!      ＼(^-^ )/
This behavior is very reasonable and is one of the functions of my favorite.

In this time, you mention the multiple call of localize() **without removing DOM**,
but I think this is no longer [prepend] function task. This is new feature at least for old i18next users.
You make a mistake in how to use.

> even though it might generate duplicate id's it won't have any problem because lib tries to find the element with id on the respective node element so it will only find one correct matching element.

I can't make sense at all.
At least a javascript library should try to deal with HTML appropriately.
If the library will support #19 condition, I think some new mechanism (like I suggested) is needed.

Thanks,

**Comment by bhaskerchari** [comment]
Hello @Shin-Ogata ,

May be your are not using a SPA, it should be ok with traditional web applications or even with SPA if page reload is fine on language change but page refresh is not acceptable in my case dealing with SPA.

@jamuhl you might need to update the Read ME then that [prepend] and [append] attributes doesn't suit well for SPA since nowadays SPA are most common web apps the communities are working.

Thanks,
Bhasker

**Comment by jamuhl** [comment]
what framework you use to build your SPA? only jquery?

Creating a spa i never used jquery to do the lifting for translation? there is stuff better suited...eg. using t function inside handlebars or whatever template helper is used.

**Comment by Shin-Ogata** [comment]
Hi @bhaskerchari, @jamuhl,

I have been working on SPA project about four years. :)
Honestly I was encountered your #19 situation, then I fixed with the way proposed that I suggested.

```
<div><span data-i18n="information"></span> is wealth.</div>
```

I think the most important problem is changing the HTML structure by **localization module**.

Changing the HTML structure should be by **NOT localization module** but the other framework or template libraries.
(Angluar, React, jQuery Mobile, Backbone, Handlbars, Hogan etc.)

Therefore, I think it's important to consider how there should be jquery-i18next as a library.
It is unrelated whether I'm conversant with SPA.

Thanks,
